src/plugins/intel_gpu/src/plugin/program.cpp

```diff
@@ -112,26 +113,181 @@ Program::Program(InferenceEngine::CNNNetwork& network, std::shared_ptr<cldnn::en
 
     auto ops = func->get_ordered_ops();
 
-    if (m_config.max_dynamic_batch > 1) {
-        // check topology for applicability
-        if (!CanProcessDynBatch(ops, networkInputs)) {
-            IE_THROW() << "Such topology cannot be compiled for dynamic batch!";
+    auto getParamName = [](const std::shared_ptr<ngraph::Node> param) -> std::string {
+        const auto& names = param->get_output_tensor(0).get_names();
+        if (!names.empty())
+            return *names.begin();
+        else
+            return param->get_friendly_name();
+    };
+
+    std::map<std::string, ngraph::PartialShape> shapes;
+    std::map<std::string, std::pair<int64_t, int64_t>> batch_dim;
+    if (m_config.enableDynamicBatch) {
+        auto param = func->get_parameters().front();
+        auto pname = getParamName(param);
+        shapes[pname] = param->get_output_partial_shape(0);
+        batch_dim[pname].first = 0;
+        batch_dim[pname].second = m_config.max_dynamic_batch;
+    } else {
+        for (const auto& param : func->get_parameters()) {
+            if (param->get_output_partial_shape(0).rank().is_dynamic()) {
+                IE_THROW() << "Dynamic rank tensors are not supported!";
+            }
+            auto layout = param->get_layout();
+            ov::PartialShape pshape = param->get_output_partial_shape(0);
+
+            //detect dynamic batch only case
+            int dynCount = 0;
+            int64_t batch_idx = -1;
+
+            for (size_t i = 0; i < pshape.size(); i++) {
+                if (pshape[i].is_dynamic()) {
+                    dynCount++;
+                    if (batch_idx < 0) {
+                        batch_idx = i;
+                    }
+                }
+            }
+
+            auto pname = getParamName(param);
+            batch_dim[pname] = {-1, -1};
+            switch (dynCount) {
+            case 2:
+                IE_THROW() << "Only dynamic batch is supported!";
+                break;
+            case 1:
+                batch_dim[pname].first = batch_idx;
+                batch_dim[pname].second = pshape[batch_idx].get_max_length();
+                pshape[batch_idx] = 1;
+            case 0:
+                shapes[pname] = pshape;
+                break;
```

Comment:
May be it makes sence to add default branch + exception msg for cases > 2?

Reply:
Fixed